src/test-util.ts

```diff
@@ -32,3 +34,195 @@ export function b(
+  const {tabId, path} = event.data;
+
+  ['error', 'warn', 'info', 'debug', 'log'].forEach(level => {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore No index signature of type 'string' was found on 'Console'.
+    console[level] = (...data: unknown[]) => {
+      window.opener.postMessage(
+        {tabId, id: 'log', result: {level, message: data.join(' ')}},
+        window.origin,
+      );
+    };
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  const AsyncFunction = Object.getPrototypeOf(async function () {}).constructor;
+  let module: typeof import('./test-util');
+  if (path !== undefined) {
+    try {
+      module = await import(window.origin + '/' + path);
+    } catch (e) {
+      console.log('Error importing ' + path + ':', e);
+    }
+  }
+
+  const handle = async (event: MessageEvent) => {
+    if (event.source === null || event.source != window.opener) {
+      return;
+    }
+    const {id, expr} = event.data;
+    const response: Response = {tabId, id};
+    if (expr) {
+      try {
+        response.result = await AsyncFunction(
+          'module',
+          '"use strict"; Object.assign(self, module); {' + expr + '}',
+        )(module);
+      } catch (e) {
+        response.error = e;
+      }
+    }
+    event.source.postMessage(response);
+  };
+
+  window.addEventListener('message', handle, false);
+  void handle(event);
+}
+
+export type Tab = {
+  run: (expr: string) => Promise<unknown>;
+  close: () => void;
+};
+
+// Registered callbacks for listener(), by tabId and call id.
+const callbacks: Record<
+  string,
+  Record<
+    string,
+    {
+      resolve: (res: unknown) => void;
+      reject?: (err: unknown) => void;
+    }
+  >
+> = {};
+
+function listener(event: MessageEvent<Response>): void {
+  const {tabId, id, result, error} = event.data;
+  const tabCallbacks = callbacks[tabId];
+  if (tabCallbacks === undefined) {
+    return;
+  }
+  if (id == 'log') {
+    const levels: Record<string, typeof console.log> = {
+      error: console.error,
+      warn: console.warn,
+      info: console.info,
+      debug: console.debug,
+    };
+    const {level, message} = result as LogResponse;
+    (levels[level] ?? console.log)('Page:', message);
+  } else {
+    const {resolve, reject} = tabCallbacks[id];
+    if (error !== undefined && reject !== undefined) {
+      reject(error);
+    } else {
+      resolve(result);
+    }
+  }
+}
+
+export async function newTab(filename?: string): Promise<Tab> {
+  const tabId = uuid();
+  const {promise, resolve} = resolver<unknown>();
+
+  if (!Object.keys(callbacks).length) {
+    window.addEventListener('message', listener, false);
+  }
+  callbacks[tabId] = {init: {resolve}};
+  const tabCallbacks = callbacks[tabId];
+
+  // Open a new tab to a URL that doesn't run anything on load.
+  const tab = window.open(
+    document.location.origin + '/src/test-util.ts',
```

Comment:
This loads the ts file as text. Probably not very meaningful. You can open `''` and you have communicate between these windows syncronously.

Reply:
It's the same domain though, allowing the use of `localStorage`! I didn't want to use the root page on this domain because that auto-runs the tests.

Reply:
`''` is special